p2p: DATA RACE in transport_mconn

```
==================
WARNING: DATA RACE
Read at 0x00c0000b03f0 by goroutine 58:
  github.com/tendermint/tendermint/p2p.newMConnConnection()
      /home/runner/work/tendermint/tendermint/p2p/transport_mconn.go:515 +0x1010
  github.com/tendermint/tendermint/p2p.(*MConnTransport).accept.func1()
      /home/runner/work/tendermint/tendermint/p2p/transport_mconn.go:198 +0x3f0

Previous write at 0x00c0000b03f0 by goroutine 36:
  github.com/tendermint/tendermint/p2p.(*MConnTransport).SetChannelDescriptors()
      /home/runner/work/tendermint/tendermint/p2p/transport_mconn.go:133 +0x4a
  github.com/tendermint/tendermint/p2p.(*Switch).OnStart()
      /home/runner/work/tendermint/tendermint/p2p/switch.go:233 +0xc2
  github.com/tendermint/tendermint/libs/service.(*BaseService).Start()
      /home/runner/work/tendermint/tendermint/libs/service/service.go:140 +0x538
  github.com/tendermint/tendermint/p2p/pex.TestPEXReactorRunning()
      /home/runner/work/tendermint/tendermint/p2p/pex/pex_reactor_test.go:108 +0x428
  testing.tRunner()
      /opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:1123 +0x202

Goroutine 58 (running) created at:
  github.com/tendermint/tendermint/p2p.(*MConnTransport).accept()
      /home/runner/work/tendermint/tendermint/p2p/transport_mconn.go:185 +0x75

Goroutine 36 (running) created at:
  testing.(*T).Run()
      /opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:1168 +0x5bb
  testing.runTests.func1()
      /opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:1439 +0xa6
  testing.tRunner()
      /opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:1123 +0x202
  testing.runTests()
      /opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:1437 +0x612
  testing.(*M).Run()
      /opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:1345 +0x3b3
  main.main()
      _testmain.go:167 +0x356
==================
--- FAIL: TestPEXReactorRunning (0.29s)
    testing.go:1038: race detected during execution of test
FAIL
```

https://github.com/tendermint/tendermint/pull/5797/checks?check_run_id=1575699557

--- a/p2p/test_util.go
+++ b/p2p/test_util.go
@@ -169,14 +169,14 @@ func MakeSwitch(
 	logger := log.TestingLogger().With("switch", i)
 	t := NewMConnTransport(logger, nodeInfo, nodeKey.PrivKey, MConnConfig(cfg))
 
-	if err := t.Listen(addr.Endpoint()); err != nil {
-		panic(err)
-	}
-
 	// TODO: let the config be passed in?
 	sw := initSwitch(i, NewSwitch(cfg, t, opts...))
 	sw.SetLogger(log.TestingLogger().With("switch", i))
 	sw.SetNodeKey(nodeKey)
+
+	if err := t.Listen(addr.Endpoint()); err != nil {
+		panic(err)
+	}
 
 	ni := nodeInfo
 	ni.Channels = []byte{}

--- a/p2p/transport_mconn.go
+++ b/p2p/transport_mconn.go
@@ -129,6 +129,11 @@ func NewMConnTransport(
 }
 
 // SetChannelDescriptors implements Transport.
+//
+// This is not concurrency-safe, and must be called before listening.
+//
+// FIXME: This is here for compatibility with existing switch code,
+// it should be passed via the constructor instead.
 func (m *MConnTransport) SetChannelDescriptors(chDescs []*conn.ChannelDescriptor) {
 	m.channelDescs = chDescs
 }

